pkg/controller/constrainttemplate/constrainttemplate_controller.go

```diff
@@ -848,3 +737,174 @@ func v1beta1ToV1(v1beta1Obj *admissionregistrationv1beta1.ValidatingAdmissionPol
 
 	return obj, nil
 }
+
+func (r *ReconcileConstraintTemplate) generateCRD(ctx context.Context, ct *v1beta1.ConstraintTemplate, proposedCRD, currentCRD *apiextensionsv1.CustomResourceDefinition, status *statusv1beta1.ConstraintTemplatePodStatus, logger logr.Logger, generateVAP bool) error {
+	if !operations.IsAssigned(operations.Generate) {
+		return nil
+	}
+	var newCRD *apiextensionsv1.CustomResourceDefinition
+	if currentCRD == nil {
+		newCRD = proposedCRD.DeepCopy()
+	} else {
+		newCRD = currentCRD.DeepCopy()
+		newCRD.Spec = proposedCRD.Spec
+	}
+
+	if err := controllerutil.SetControllerReference(ct, newCRD, r.scheme); err != nil {
+		return err
+	}
+
+	if currentCRD == nil {
+		logger.Info("creating crd")
+		if err := r.Create(ctx, newCRD); err != nil {
+			err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not create CRD", status, err)
+			return err
+		}
+	} else if !reflect.DeepEqual(newCRD, currentCRD) {
+		logger.Info("updating crd")
+		if err := r.Update(ctx, newCRD); err != nil {
+			err := r.reportErrorOnCTStatus(ctx, ErrUpdateCode, "Could not update CRD", status, err)
+			return err
+		}
+		if ct.Annotations != nil {
+			ct.Annotations[constraint.BlockVAPBGenerationUntilAnnotation] = ""
+		}
+	}
+	if !generateVAP {
+		return nil
+	}
+
+	err := r.updateTemplateWithBlockVAPBGenerationAnnotations(ctx, ct)
```

Comment:
Should probably note that we are purposefully making a second call to the API server here in order to make sure the timestamp is post-CRD creation.

Otherwise the question I have is "why make two requests?"

Reply:
Added a comment here.